File: 0526/내려가기.py
import sys
input = sys.stdin.readline

N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]

dp_min = [e for e in arr[0]]
dp_max = [e for e in arr[0]]

if N == 1:
    print(max(arr[0]), min(arr[0]))

else:
    i = 1
    while i < N:
        x = [0] * 3
        y = [0] * 3
        for j in range(3):
            if j == 0:
                x[j] = min(dp_min[j] + arr[i][j], dp_min[j + 1] + arr[i][j])
                y[j] = max(dp_max[j] + arr[i][j], dp_max[j + 1] + arr[i][j])
            elif j == 2:
                x[j] = min(dp_min[j] + arr[i][j], dp_min[j - 1] + arr[i][j])
                y[j] = max(dp_max[j] + arr[i][j], dp_max[j - 1] + arr[i][j])
            else:
                x[j] = min(dp_min[j] + arr[i][j], dp_min[j + 1] + arr[i][j], dp_min[j - 1] + arr[i][j])
                y[j] = max(dp_max[j] + arr[i][j], dp_max[j + 1] + arr[i][j], dp_max[j - 1] + arr[i][j])

        i += 1

        for j in range(3):
            dp_min[j] = x[j]
            dp_max[j] = y[j]

    print(max(dp_max), min(dp_min))

File: 0526/내려가기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

    static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    static int N;
    static int[][] arr;
    static int[][] maxDp;
    static int[][] minDp;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(bf.readLine());
        arr = new int[N][3];
        maxDp = new int[N][3];
        minDp = new int[N][3];

        for (int i=0; i<N; i++) {
            String input = bf.readLine();
            String[] split = input.split(" ");
            arr[i][0] = Integer.parseInt(split[0]);
            arr[i][1] = Integer.parseInt(split[1]);
            arr[i][2] = Integer.parseInt(split[2]);
        }

        maxDp[0][0] = arr[0][0];
        maxDp[0][1] = arr[0][1];
        maxDp[0][2] = arr[0][2];

        minDp[0][0] = arr[0][0];
        minDp[0][1] = arr[0][1];
        minDp[0][2] = arr[0][2];

        for (int i=1; i<N; i++) {
            maxDp[i][0] = Math.max(maxDp[i - 1][0], maxDp[i - 1][1]) + arr[i][0];
            maxDp[i][1] = Math.max(maxDp[i - 1][0], Math.max(maxDp[i - 1][1], maxDp[i - 1][2])) + arr[i][1];
            maxDp[i][2] = Math.max(maxDp[i - 1][1], maxDp[i - 1][2]) + arr[i][2];

            minDp[i][0] = Math.min(minDp[i - 1][0], minDp[i - 1][1]) + arr[i][0];
            minDp[i][1] = Math.min(minDp[i - 1][0], Math.min(minDp[i - 1][1], minDp[i - 1][2])) + arr[i][1];
            minDp[i][2] = Math.min(minDp[i - 1][1], minDp[i - 1][2]) + arr[i][2];
        }

        System.out.println(Arrays.stream(maxDp[N - 1]).max().getAsInt() + " " + Arrays.stream(minDp[N - 1]).min().getAsInt());
    }
}
